### ISSUE TITLE
Rewrite audio system

### ISSUE THREAD

**Issue by Krzmbrzl** [mumble-voip/mumble#5408]
### Context

The currently used audio system in the client is rather messy and has quite a few quirks that can't be easily fixed.
On top of that the code itself is written in a pretty bad style that makes it rather hard to grasp what is going on.

### Description

For this reason, the audio handling should be rewritten from scratch and while doing so, special attention should be paid to fixing the currently experienced audio issues.

### Mumble component

Client

### OS-specific?

No

### Additional information

The current issues that need to be fixed are:
- #4883: Audio stopping to work when applying settings (only on some machines)
- #4177, #1092: Audio system consuming a lot of CPU
- #5314, #1089: Stop/Reduce audio processing when muted - see also implementation attempt #171 and the revert in #4633
- #5379: Close input sinks when user is deafened (allows e.g. Bluetooth protocol to switch)
- #3491 (especially https://github.com/mumble-voip/mumble/issues/3491#issuecomment-606189582): Try to use a different Bluetooth protocol (in some cases) to allow better audio quality

On the code side, we should
- get rid of `goto` statements
- split processing into smaller functions
- Use smart-pointers
- Use more descriptive variable names
- Don't use global variables (that much)
- Don't use magic & hardcoded numbers; prefer named constants

**Comment by vimpostor** [comment]
Also a rather controversial design decision is that the Audio backend can be different for output and input. Is there actually any usecase for this? This does not only complicate the code but it also is unintuitive for the user, as you have to change the backend in the settings twice (unless using the Audio wizard).
Is there anyone that actually uses different backends? Otherwise I would vote to remove this implementation detail, which would also allow to make other things easier such as registering Mumble as a single node (with input and output ports) in the Pipewire/Jack node graph, instead of registering two nodes with one having the input ports and the other the output ports.

**Comment by TerryGeng** [comment]
I agree with @vimpostor 's idea. In fact, the separation of the backend for input and output is not very compatible with macOS, especially when it comes to the native echo cancellation of Mac...

**Comment by davidebeatrici** [comment]
The backend separation was extremely useful on Linux before PipeWire became a thing: it was not rare to see JACK as input and PulseAudio as output.

Right now I can only think of Windows with WASAPI + ASIO as a reason for the feature to exist, but as far as I know it's not an ideal setup anymore now that WASAPI itself can deliver low latency (related: #3503).

**Comment by AtmanActive** [comment]
One extremely useful feature for me is the ability to use two distinct audio devices via WASAPI, one for input, the other for output, each with it's own bitrate/depth and have Mumble work flawlessly with it.

**Comment by Krzmbrzl** [comment]
Relevant: https://github.com/mumble-voip/libcrossaudio
